### ISSUE TITLE
Reading data with queue is even slower than using feed_dict

### ISSUE THREAD

**Issue by akaitsuki-ii** [tensorflow/tensorflow#7817]
It seems reading data with feed_dict is inefficient in tensorflow (I found it 3-4 times slower than a theano-based implementation with a totally same network structure), I turn to a queue-based method as official recommendation. However, my experiment result give a even worse performance. Is there anything wrong ?

### Related issue
[#3377 Moving data from CPU to GPU is slow](https://github.com/tensorflow/tensorflow/issues/3377)

### Environment info
Operating System: Ubuntu 16.04
CPU: Intel i4790-k
GPU: Nvidia GTX 1070 (8 GB)
Memory: 16 GB
Tensorflow version: 1.0
CUDA version: 8.0
cuDNN version: 5.0

### Implementation Example
1) Using feed_dict
```python
x, y = np.array(...)
in_x, in_y = tf.placeholder(...)
""" f() is some computation in the network, including embedding_lookup, 
bidirectional_dynamic_rnn, dense """
train_op = f(in_x, in_y) 
sess = tf.Session()
for _ in range(num_epochs): # num_epochs is 100 here
    sess.run(train_op, {in_x: x, in_y: y})
```
2) Using queue
```python
x, y = np.array(...)
x, y = tf.convert_to_tensor(...)
in_x, in_y = tf.train.slice_input_producer([x, y], num_epochs=100)
in_x, in_y = tf.train.batch([in_x, in_y], batch_size=32)
train_op = f(in_x, in_y)
sess = tf.Session()
coord = tf.train.Coordinator()
threads = tf.train.start_queue_runners(sess, coord)
try:
    while not coord.should_stop():
        sess.run(train_op)
    except Exception as e:
        coord.request_stop(e)
    finally:
        coord.request_stop()
coord.join(threads)
```

**Comment by yaroslavvb** [comment]
I troubleshooted similar issue [here](http://stackoverflow.com/questions/39840323/benchmark-of-howto-reading-data/39842628#39842628) and it was caused by Python thread scheduler choosing a bad strategy. Essentially Python would schedule computation thread that issues a single enqueue call, then this thread would block and have to be pre-empted. Python doesn't support parallel execution of Python code and pre-emption is slow so this part is a performance hit. Eventually it pre-empts dequeue thread to schedule enqueue thread, which does a single enqueue call before Python decides to give execution back to main thread. This back-and-forth dance introduced 10x slowdown in the pipeline.

One solution is to make sure that you never have queue starvation, for instance, by making enqueue part faster (ie by using enqueue many as here https://github.com/tensorflow/tensorflow/issues/3009 ) and by making dequeue part slower (ie, add more computation to `train_op` until it becomes slow enough to be bottleneck)

**Comment by akaitsuki-ii** [comment]
@yaroslavvb Thx for your advice.
However, I fail to fix the problem. I try replace *tf.train.batch* with *tf.train.shuffle_batch* and set *min_after_dequeue=10000*  to avoid queue starvation (I assume this parameter help fix it, right ?), but my program as slow as before.
What's more, I follow the [link](https://github.com/tensorflow/tensorflow/issues/1824#issuecomment-225754659) that you mentioned to generate timeline. It did take a long time to generate a timeline.json file (600 MB+, with correct .json file format), but failed to be visualized using *chrome://tracing* and give a blank output. Do you have any idea of what's going wrong ?

**Comment by yaroslavvb** [comment]
What I meant was -- make sure your `dequeue` part is slower than `enqueue`. Pre-loading the queue is not really a solution because if your computation runs faster than enqueueing, it'll consume all the elements on the queue and you'll be back to queue starvation. The reality is that queues are bad when your computation is extremely cheap because of the way Python multi-threading works, feed_dict may actually be better for those cases.

**Comment by akaitsuki-ii** [comment]
@yaroslavvb Thank you for reminding me of checking in TensorBoard.
I do some test and it seems that *enqueue* part is not a bottleneck of the program, it's full (and capacity is large enough) all the time. There is no problem of starvation.
And I find that *dequeue* operation really cause problems, according to the compute time statistics in TensorBoard, *dequeue* cost 6x compute time than my model, which result in low GPU-Util and even worse performance than using feed_dict.

**Comment by yaroslavvb** [comment]
You could narrow it down by making your program single threaded - first
preload the queue and then consume it in the same thread. If it's still
slow, then could narrow it down to a case that spends most time in dequeue
and is single threaded, at which point CPU profile can tell where the time
is spent

**Comment by yaroslavvb** [comment]
BTW, here's an example of making reader pipeline single-threaded -- https://github.com/yaroslavvb/stuff/blob/master/ericyue-slowreader/benchmark-batch.py

Especially see the line with `sess.run(a_queue_qr.enqueue_ops)`. Restructuring the problem in this way can help rule out some Python-threads related strangeness going on

**Comment by akaitsuki-ii** [comment]
@yaroslavvb Sorry for check it so late.
I did preload the queue full enough (using single thread `sess.run(enqueue_ops)`) this time before start running my model and consume the queue, there is no more enqueue operation after dequeue is start.
But it seems to be very slow for dequeue in this situation.

The following is the graph and node metadata showed in TensorBoard. You can see that compute time of dequeue operation is 24.2ms, which is much longer than that of my model, 1.36ms.
![graph](https://cloud.githubusercontent.com/assets/8386553/23578611/322056d4-0116-11e7-9f48-621e7b6a55d0.png)
![dequeue-status](https://cloud.githubusercontent.com/assets/8386553/23578630/a4d7c284-0116-11e7-81f4-7d4aec8b49cd.png)
![model-status](https://cloud.githubusercontent.com/assets/8386553/23578632/aecdc41e-0116-11e7-8047-501c6e3dd442.png)

However, if I replace my model with much light operation like `tf.square`, dequeue will be much faster with same tensors output.
![graph-light](https://cloud.githubusercontent.com/assets/8386553/23578663/45ab84e8-0117-11e7-9d6f-7c6316857d2f.png)
![dequeue-status-light](https://cloud.githubusercontent.com/assets/8386553/23578664/494bfe02-0117-11e7-94e0-a43dccceb613.png)

I can not figure out what the problem is...?

**Comment by akaitsuki-ii** [comment]
@yaroslavvb 
My implementation rely on 3rd library tensorlayer and it takes me sometime to rewrite it using tensorflow only.
You can check and download the example .py file with [this](https://www.dropbox.com/sh/ijc7lccquf1g4l7/AABX9pf5LIqkvPxplbhFRVafa?dl=0) Dropbox link.
Please check different performance of *dequeue* operation between `sess.run(train_op)` and `sess.run(light_op)`.

**Comment by yaroslavvb** [comment]
@akaitsuki-ii so I tried making a timeline for your stuff (https://github.com/yaroslavvb/stuff/commit/5d2feffd62fa50d71dd2dfd314ee405c21d58a25), and I couldn't because your execution graph is too large. There are 942,316 ops being executed in the first session.run call. The dump of stepstats is over a GB and timeline generation chokes. So I think that's the real problem, and the "enqueue" op being slow is a false positive since tensorboard couldn't parse all the ops

**Comment by akaitsuki-ii** [comment]
@yaroslavvb
Errr...In fact I did try to make timeline before and fail because of the same reason.
There may be a mistake that do you really mean that *dequeue* op being slow is false positive? As *enqueue* op is not contained in `sess.run` call.
But if *dequeue* op is not really the problem, why I get worse performance changing to queue-based implementation than feeding with feed_dict at each `sess.run` call ?
Is it just better using feed_dict instead of queue_based method in this situation ?

**Comment by yaroslavvb** [comment]
There are about 1 million ops being executed (`grep node_stats stepstats-1.json | wc`), and it takes 4.72 second on my 2014 MacBook, which seems fairly fast for 1 million ops. Such large computation graphs are outside of typical range of cases TensorFlow is tuned for, so there's significant overhead involved.

As to why I said it is a false positive, if you open `stepstats-1.json` generated by the commit I referenced, you will see the following

```
    node_stats {
      node_name: "random_shuffle_queue_DequeueMany/n"
      all_start_micros: 1489378683948609
      op_start_rel_micros: 1
      op_end_rel_micros: 1
      all_end_rel_micros: 52
      memory {
        allocator_name: "cpu"
```

So the Dequeue op takes 51 microseconds.


As to why `feed_dict` is faster, it's an interesting question, and would require some digging to getto. One theory is that `dequeue` op requires a kernel launch, so when you already have a million ops being launched, perhaps the thread scheduling system is overloaded,  any new kernel launches are slow. Meanwhile, `feed_dict` approach does not add an additional op for reading data from queue, so no kernel launch.

**Comment by alextp** [comment]
I'm not sure I follow the explanation for feed_dict, @yaroslavvb , since feed_dict translates to adding a Recv op in the graph for each thing being fed, and those still have to run.

Dequeue on the other hand is always suspicious since it can be sometimes fast and sometimes slow (if the queue is often full but sometimes empty). It might make sense to aggregate the costs across many invocations before saying it's not the issue.

**Comment by yaroslavvb** [comment]
@alextp Good point about feed_dict, I can [see](https://github.com/yaroslavvb/stuff/blob/master/akaitsuki-slow/feed_dict.pbtxt) an extra _Recv node in that graph.

BTW that script removes that queue variability -- there's no queue runners, things are first loaded onto the queue, and then consumed from the queue in the same thread.

**Comment by hanxiao** [comment]
Just want to add something here, I implemented a multiprocess-based data feeding pipeline for multi-task learning. avg. GPU utilization >90% and quad-core CPU utilization >95%. In case anyone interested: https://hanxiao.github.io/2017/07/07/Get-10x-Speedup-in-Tensorflow-Multi-Task-Learning-using-Python-Multiprocessing/

**Comment by ghamarian** [comment]
I don't know if it is still relevant, but reading through the discussion I do not find the reason or the solution to the problem. I have a very simple network (no operations almost) and I use:
tf.train.shuffle_batch(decoded, capacity=batch_size * 50,
                                 batch_size=batch_size,
                                 num_threads=6,
                                 min_after_dequeue=batch_size * 10,
                                 allow_smaller_final_batch=True)

As my performance drops as I increase the batch_size almost linearly, i.e., increasing batch_size 4 times, it takes 4 times as much.
